src/Coordination/KeeperStateMachine.cpp

```diff
@@ -150,48 +151,88 @@ void assertDigest(
 nuraft::ptr<nuraft::buffer> KeeperStateMachine::pre_commit(uint64_t log_idx, nuraft::buffer & data)
 {
     auto request_for_session = parseRequest(data);
-    if (!request_for_session.zxid)
-        request_for_session.zxid = log_idx;
+    if (!request_for_session->zxid)
+        request_for_session->zxid = log_idx;
 
-    preprocess(request_for_session);
+    preprocess(*request_for_session);
     return nullptr;
 }
 
-KeeperStorage::RequestForSession KeeperStateMachine::parseRequest(nuraft::buffer & data)
+std::shared_ptr<KeeperStorage::RequestForSession> KeeperStateMachine::parseRequest(nuraft::buffer & data, bool final)
 {
     ReadBufferFromNuraftBuffer buffer(data);
-    KeeperStorage::RequestForSession request_for_session;
-    readIntBinary(request_for_session.session_id, buffer);
+    auto request_for_session = std::make_shared<KeeperStorage::RequestForSession>();
+    readIntBinary(request_for_session->session_id, buffer);
 
     int32_t length;
     Coordination::read(length, buffer);
 
     int32_t xid;
     Coordination::read(xid, buffer);
 
+    static constexpr std::array non_cacheable_xids{
+        Coordination::WATCH_XID,
+        Coordination::PING_XID,
+        Coordination::AUTH_XID,
+        Coordination::CLOSE_XID,
+    };
+
+    const bool should_cache = request_for_session->session_id != -1 && data.size() > min_request_size_to_cache
+        && std::all_of(non_cacheable_xids.begin(),
+                       non_cacheable_xids.end(),
+                       [&](const auto non_cacheable_xid) { return xid != non_cacheable_xid; });
+
+    if (should_cache)
+    {
+        std::lock_guard lock(request_cache_mutex);
+        if (auto xid_to_request_it = parsed_request_cache.find(request_for_session->session_id);
+            xid_to_request_it != parsed_request_cache.end())
+        {
+            auto & xid_to_request = xid_to_request_it->second;
+            if (auto request_it = xid_to_request.find(xid); request_it != xid_to_request.end())
```

Comment:
>  Also looks like we remove requests from cache after commit.

Yes, because we will not use the request after that point (or after rollback). Added a comment saying just that.

> Need to check is wrong/non-unique XID is a protocol violation or not?

XID isn't mentioned a lot but based on ZK implementation, it's atomically increased on each request inside a session.
Also, we already rely on this fact for processing read requests.

We know for a fact what our client does, and we can disable the parse cache by setting `min_request_size_for_cache` to 0